runtime-sdk/modules/evm/src/precompile/confidential.rs

```diff
@@ -165,14 +207,243 @@ pub(super) fn call_deoxysii_open(
+    let (seed_len, input) = utils::cut_uint_from_params(input, "seed length out of bounds".into())?;
+
+    if input.len() < utils::round_up(seed_len) {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other("input too short for seed".into()),
+        });
+    }
+
+    let seed = &input[..seed_len];
+
+    let sig_type: SignatureType = <usize as TryInto<u8>>::try_into(method)
+        .map_err(|_| PrecompileFailure::Error {
+            exit_status: ExitError::Other("method identifier out of bounds".into()),
+        })?
+        .try_into()
+        .map_err(|_| PrecompileFailure::Error {
+            exit_status: ExitError::Other("unknown signature type".into()),
+        })?;
+    let signer = signature::MemorySigner::new_from_seed(sig_type, seed).map_err(|err| {
+        PrecompileFailure::Error {
+            exit_status: ExitError::Other(format!("error creating signer: {}", err).into()),
+        }
+    })?;
+    let public = signer.public_key().as_bytes().to_vec();
+    let private = signer.to_bytes();
+
+    let gas_cost = linear_cost(
+        target_gas,
+        input.len() as u64,
+        *KEYPAIR_GENERATE_BASE_COST
+            .lock()
+            .unwrap()
+            .get(&sig_type)
+            .unwrap(),
+        0,
+    )?;
+
+    let mut output: Vec<u8> = Vec::new();
+    utils::push_uint_to_params(&mut output, public.len());
+    utils::push_uint_to_params(&mut output, private.len());
+    utils::push_slice_to_params(&mut output, &public, true);
+    utils::push_slice_to_params(&mut output, &private, true);
+    Ok(PrecompileOutput {
+        exit_status: ExitSucceed::Returned,
+        cost: gas_cost,
+        output,
+        logs: Default::default(),
+    })
+}
+
+pub(super) fn call_sign_message_with_context(
+    input: &[u8],
+    target_gas: Option<u64>,
+    _context: &Context,
+    _is_static: bool,
+) -> PrecompileResult {
+    // 4 fixed parameters, followed by three strings.
+    const SLOTS: usize = 4;
+
+    // Input encoding: uint method || uint pk_len || uint context_len || uint message_len || pk || context || message.
+    // Only the bottom byte of the method is taken into account.
+    if input.len() < SLOTS * WORD {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(
+                format!("input length must be at least {} bytes", SLOTS * WORD).into(),
+            ),
+        });
+    }
+
+    let (method, input) =
+        utils::cut_uint_from_params(input, "method identifier out of bounds".into())?;
+    let (pk_len, input) =
+        utils::cut_uint_from_params(input, "private key length out of bounds".into())?;
+    let (context_len, input) =
+        utils::cut_uint_from_params(input, "context length out of bounds".into())?;
+    let (message_len, input) =
+        utils::cut_uint_from_params(input, "message length out of bounds".into())?;
```

Comment:
I thought we were going to sign prehashed except in the case of sr25519?

Reply:
The way these crates are made, there's no way to just give them the bytes of an existing hash. I tried going the dummy hasher route, but that proved impossible too, so the current code has multiple combinations of signer+hash and always hashes the message itself.